### ISSUE TITLE
Support ordered-set (WITHIN GROUP) aggregates.

### ISSUE THREAD

**Issue by sim1984** [FirebirdSQL/firebird#7632]
Need to add support for ordered-set and hypothetical-set aggregate functions, as well as implementations of the instances defined in SQL:2008 (percentile_cont(), percentile_disc(), rank(), dense_rank(), percent_rank(), cume_dist()).  It is good to added mode() though it is not in the spec. In addition, you can gracefully solve the problem of element unordering in the LIST aggregate function (in the SQL standard LISTAGG).

Ordered-Set Aggregate Functions
```
PERCENTILE_CONT( <expr> ) WITHIN GROUP ( ORDER BY <order_expr> )
PERCENTILE_DISC( <expr> ) WITHIN GROUP ( ORDER BY <order_expr> )
LISTAGG( <expr>, <separator> ) WITHIN GROUP ( ORDER BY <order_expr> )
```

It is also possible to introduce the LISTAGG function, and make LIST an alias for compatibility with the standard.

Hypothetical-Set Aggregate Functions
```
RANK( <args> ) WITHIN GROUP ( ORDER BY <order_exprs> )
DENSE_RANK( <args> ) WITHIN GROUP ( ORDER BY <order_exprs> )
PERCENT_RANK( <args> ) WITHIN GROUP ( ORDER BY <order_exprs> )
CUME_DIST( <expr> ) WITHIN GROUP ( ORDER BY <order_exprs> )
```

You can read more about how this is done in PostgreSQL here.
https://www.postgresql.org/docs/current/functions-aggregate.html
https://paquier.xyz/postgresql-2/postgres-9-4-feature-highlight-within-group/

**Comment by dyemanov** [comment]
If `LISTAGG` is added for standard compatibility, should legacy `LIST` also support ordering? If yes, should it follow the standard syntax (so that `LIST` and `LISTAGG` just become synonyms) or, being non-standard, may use its own syntax for ordring, e.g. `LIST(FIELD1 ORDER BY FIELD2)`?

**Comment by sim1984** [comment]
I think it's enough to make it just a synonym. But if they make their own version to simplify, then do not  object.

**Comment by dyemanov** [comment]
The standard `LISTAGG` also has the separator string being mandatory while our `LIST` implementation has it optional, with a comma implied if omitted. And the standard `LISTAGG` defines the separator as a literal only, while we allow arbitrary value expressions. Would it be OK to preserve these differences for the `LISTAGG` too?

**Comment by dyemanov** [comment]
Sorting may be mandatory in `LISTAGG` (as per SQL standard), but it must remain optional (or missing completely) in `LIST` due to backward compatibility.

**Comment by dyemanov** [comment]
Oracle supports `LISTAGG` is mostly standard way, but it has optional separator. So maybe we should preserve this syntax.

**Comment by sim1984** [comment]
Agree. In this case, we have full compatibility with the current LIST function, with the exception of the mandatory sorting. For LIST, sorting can be left optional for backward compatibility reasons. Please note in the documentation that in new projects it is recommended to use the standard LISTAGG.

**Comment by sim1984** [comment]
`PERCENTILE_CONT` and `PERCENTILE_DISC` functions implimented in https://github.com/FirebirdSQL/firebird/pull/8946

**Comment by mrotteveel** [comment]
Is this feature complete or not?

**Comment by sim1984** [comment]
> Is this feature complete or not?

PERCENTILE_CONT and PERCENTILE_DISC have some minor drawbacks, but they're not critical. (There is no argument constancy check within the group when using these functions as window functions)
But Hypothetical-Set Aggregate Functions aren't implemented yet. I'm not sure we need them.

**Comment by mrotteveel** [comment]
@sim1984 I meant the ordered-set feature (WITHIN GROUP) itself.

**Comment by sim1984** [comment]
@mrotteveel
As I already mentioned, the PERCENTILE_CONT and PERCENTILE_DISC functions are implemented. The main drawback of the implementation is the lack of a check that the argument (percentile) of the PERCENTILE_CONT/PERCENTILE_DISC functions must be constant within a group. This check exists when PERCENTILE_CONT/PERCENTILE_DISC are used as aggregate functions, but there is no check when PERCENTILE_CONT/PERCENTILE_DISC are used as window functions.

I'll give an example of what I mean.

For aggregates (Good):
```
WITH
  T(N, P) AS (
     SELECT 1, 0.5 FROM RDB$DATABASE
     UNION ALL
     SELECT 2, 0.5 FROM RDB$DATABASE
     UNION ALL
     SELECT 3, 1 FROM RDB$DATABASE
  )
SELECT
  PERCENTILE_DISC(0.5) WITHIN GROUP(ORDER BY N) 
FROM T;

PERCENTILE_DISC
===============
              2

WITH
  T(N, P) AS (
     SELECT 1, 0.5 FROM RDB$DATABASE
     UNION ALL
     SELECT 2, 0.5 FROM RDB$DATABASE
     UNION ALL
     SELECT 3, 1 FROM RDB$DATABASE
  )
SELECT
  PERCENTILE_DISC(P) WITHIN GROUP(ORDER BY N) 
FROM T;

Statement failed, SQLSTATE = 42000
Dynamic SQL Error
-SQL error code = -104
-Argument for PERCENTILE_DISC function must be constant within each group

WITH
  T(N, P) AS (
     SELECT 1, 0.5 FROM RDB$DATABASE
     UNION ALL
     SELECT 2, 0.5 FROM RDB$DATABASE
     UNION ALL
     SELECT 3, 1 FROM RDB$DATABASE
  )
SELECT
  PERCENTILE_DISC(P) WITHIN GROUP(ORDER BY N) 
FROM T
GROUP BY P;

PERCENTILE_DISC
===============
              1
              3

```

For window functions (Bad):
```
-- OK
WITH
  T(N, P) AS (
     SELECT 1, 0.5 FROM RDB$DATABASE
     UNION ALL
     SELECT 2, 0.5 FROM RDB$DATABASE
     UNION ALL
     SELECT 3, 1 FROM RDB$DATABASE
  )
SELECT
  N, P,
  PERCENTILE_DISC(0.5) WITHIN GROUP(ORDER BY N) OVER()
FROM T;

           N                     P PERCENTILE_DISC
============ ===================== ===============
           1                   0.5               2
           2                   0.5               2
           3                   1.0               2

-- There must be an error here. You can't change the <percentile> argument inside a group.
WITH
  T(N, P) AS (
     SELECT 1, 0.5 FROM RDB$DATABASE
     UNION ALL
     SELECT 2, 0.5 FROM RDB$DATABASE
     UNION ALL
     SELECT 3, 1 FROM RDB$DATABASE
  )
SELECT
  N, P,
  PERCENTILE_DISC(P) WITHIN GROUP(ORDER BY N) OVER()
FROM T;

           N                     P PERCENTILE_DISC
============ ===================== ===============
           1                   0.5               2
           2                   0.5               2
           3                   1.0               2

-- OK
WITH
  T(N, P) AS (
     SELECT 1, 0.5 FROM RDB$DATABASE
     UNION ALL
     SELECT 2, 0.5 FROM RDB$DATABASE
     UNION ALL
     SELECT 3, 1 FROM RDB$DATABASE
  )
SELECT
  N, P,
  PERCENTILE_DISC(P) WITHIN GROUP(ORDER BY N) OVER(PARTITION BY P)
FROM T;

           N                     P PERCENTILE_DISC
============ ===================== ===============
           1                   0.5               1
           2                   0.5               1
           3                   1.0               3
```

That is, the `<percentile>` argument is only allowed to be modified if it's included in the `PARTITION BY` clause within the window. But I don't know how to check this.

I don't think this is a critical error that would cause us to delay merging this PR.

*************************************

Hypothetical-Set Aggregate Functions also use WITHIN GROUP, but they are a separate category of functions. I may implement them later.

**Comment by sim1984** [comment]
The aggregate functions of the hypothetical set are implemented in https://github.com/FirebirdSQL/firebird/pull/9035
